pkg/controller/stack/stack_controller.go

```diff
@@ -765,6 +805,119 @@ func (sess *reconcileStackSession) lookupPulumiAccessToken(ctx context.Context)
 	return "", false
 }
 
+type ProjectFile struct {
+	Name    string `json:"name"`
+	Runtime string `json:"runtime"`
+	pulumiv1.ProgramSpec
+}
+
+func (sess *reconcileStackSession) SetupWorkdirFromYAML(ctx context.Context, programRef shared.ProgramReference) (_commit string, retErr error) {
+	sess.logger.Debug("Setting up pulumi workdir for stack", "stack", sess.stack)
+
+	// Create a new workspace.
+	secretsProvider := auto.SecretsProvider(sess.stack.SecretsProvider)
+
+	// Create the temporary workdir
+	dir, err := os.MkdirTemp("", "pulumi_auto")
+	if err != nil {
+		return "", fmt.Errorf("unable to create tmp directory for workspace: %w", err)
+	}
+	sess.rootDir = dir
+
+	// Cleanup the rootdir on failure setting up the workspace.
+	defer func() {
+		if retErr != nil {
+			_ = os.RemoveAll(sess.rootDir)
+		}
+	}()
+
+	program := pulumiv1.Program{}
+	programKey := client.ObjectKey{
+		Name:      programRef.Name,
+		Namespace: sess.namespace,
+	}
+
+	err = sess.kubeClient.Get(ctx, programKey, &program)
+	if err != nil {
+		return "", fmt.Errorf("unable to retrieve program for stack: %w", err)
+	}
+
+	var project ProjectFile
+	project.Name = program.Name
+	project.Runtime = "yaml"
+	project.ProgramSpec = program.Program
+
+	out, err := yaml.Marshal(&project)
+	if err != nil {
+		return "", fmt.Errorf("failed to marshal program object to YAML: %w", err)
+	}
+
+	err = os.WriteFile(filepath.Join(dir, "Pulumi.yaml"), out, 0600)
+	if err != nil {
+		return "", fmt.Errorf("failed to write YAML to file: %w", err)
+	}
+
+	var w auto.Workspace
+	w, err = auto.NewLocalWorkspace(ctx, auto.WorkDir(dir), secretsProvider)
+	if err != nil {
+		return "", fmt.Errorf("failed to create local workspace: %w", err)
+	}
+
+	sess.workdir = w.WorkDir()
```

Comment:
NB: in #324 I have factored here down out into a helper; watch out for that if you end up rebasing.